cmd/kubeadm/app/cmd/phases/join/controlplanejoin.go

```diff
@@ -0,0 +1,256 @@
+	markcontrolplanephase "k8s.io/kubernetes/cmd/kubeadm/app/phases/markcontrolplane"
+	uploadconfigphase "k8s.io/kubernetes/cmd/kubeadm/app/phases/uploadconfig"
+	"k8s.io/kubernetes/pkg/util/normalizer"
+)
+
+var (
+	controlPlaneJoinExample = normalizer.Examples(`
+		# Generates all static Pod manifest files for control plane components,
+		# functionally equivalent to what is generated by kubeadm init.
+		kubeadm init phase control-plane all
+		# Generates all static Pod manifest files using options read from a configuration file.
+		kubeadm init phase control-plane all --config config.yaml
+		`)
+
+	controlPlanePhaseProperties = map[string]struct {
+		name  string
+		short string
+	}{
+		kubeadmconstants.KubeAPIServer: {
+			name:  "apiserver",
+			short: getPhaseDescription(kubeadmconstants.KubeAPIServer),
+		},
+		kubeadmconstants.KubeControllerManager: {
+			name:  "controller-manager",
+			short: getPhaseDescription(kubeadmconstants.KubeControllerManager),
+		},
+		kubeadmconstants.KubeScheduler: {
+			name:  "scheduler",
+			short: getPhaseDescription(kubeadmconstants.KubeScheduler),
+		},
+	}
+)
+
+type controlPlaneJoinData interface {
+	Cfg() *kubeadmapi.JoinConfiguration
+	ClientSetFromFile(string) (*clientset.Clientset, error)
+	InitCfg() (*kubeadmapi.InitConfiguration, error)
+}
+
+func getPhaseDescription(component string) string {
+	return fmt.Sprintf("Generates the %s static Pod manifest", component)
+}
+
+func getControlPlaneJoinPhaseFlags() []string {
+	return []string{
+		options.APIServerAdvertiseAddress,
+		options.APIServerBindPort,
+		options.CfgPath,
+		options.ControlPlane,
+		options.NodeName,
+		options.TokenDiscovery,
+		options.TokenDiscoveryCAHash,
+		options.TokenDiscoverySkipCAHash,
+	}
+}
+
+// NewControlPlaneJoinPhase creates a kubeadm workflow phase that implements marking the new node as master and update the cluster status
+// with information about current node
+func NewControlPlaneJoinPhase() workflow.Phase {
+	return workflow.Phase{
+		Name:  "control-plane-join",
+		Short: "Prepares the machine to join a control plane.",
+		Long:  cmdutil.MacroCommandLongDescription,
+		Phases: []workflow.Phase{
+			{
+				Name:           "all",
+				Short:          "Generates all static Pod manifest files",
+				InheritFlags:   getControlPlaneJoinPhaseFlags(),
+				RunAllSiblings: true,
+			},
+			newControlPlaneJoinSubphase(kubeadmconstants.KubeAPIServer),
+			newControlPlaneJoinSubphase(kubeadmconstants.KubeControllerManager),
+			newControlPlaneJoinSubphase(kubeadmconstants.KubeScheduler),
+			newEtcdLocalSubphase(),
+			newUploadConfigSubphase(),
+			newMarkControlPlaneSubphases(),
+		},
+		InheritFlags: getControlPlaneJoinPhaseFlags(),
+	}
+}
+
+func newControlPlaneJoinSubphase(component string) workflow.Phase {
+	return workflow.Phase{
+		Name:         controlPlanePhaseProperties[component].name,
+		Short:        controlPlanePhaseProperties[component].short,
+		Run:          runControlPlaneJoinSubphase(component),
```

Comment:
rename this to `runControlPlaneJoinComponentSubphase`

Reply:
removed because it is already implemented in control-plane-prepare phase